### PR TITLE
polish(pu): polish efficiency and performance on atari and dmc

### PR DESCRIPTION
**Pull request by puyuan1996**
- Optimize efficiency, specifically by replacing deepcopy with a predefined share_kv_cache_pool.
- Add muzero_segment_collector, enable the buffer reanalyze option, and add the train_***_segment.py entry point.
- Refine the configurations for Atari and DeepMind Control environments.

### CODE REVIEW

File: lzero/entry/train_unizero.py [hunk 1]
```diff
@@ -107,7 +107,8 @@ def train_unizero(
     batch_size = policy._cfg.batch_size
 
     # TODO: for visualize
-    # stop, reward = evaluator.eval(learner.save_checkpoint, learner.train_iter, collector.envstep)
+    stop, reward = evaluator.eval(learner.save_checkpoint, learner.train_iter, collector.envstep)
+    # import sys; sys.exit(0)
```
Comment (PaParaZz1) [inline lzero/entry/train_unizero.py:111]: remove commented code

File: lzero/model/common.py [hunk 2]
```diff
@@ -341,7 +341,11 @@ def __init__(
         self.activation = activation
         self.embedding_dim = embedding_dim
 
-        self.last_linear = nn.Linear(64 * 8 * 8, self.embedding_dim, bias=False)
+        if self.observation_shape[1] == 64:
```
Comment (PaParaZz1) [inline lzero/model/common.py:344]: add comments about different size

File: lzero/model/unizero_world_models/lpips.py [hunk 3]
```diff
@@ -21,16 +21,16 @@ def __init__(self, use_dropout: bool = True):
         self.scaling_layer = ScalingLayer()
         self.chns = [64, 128, 256, 512, 512]  # vg16 features
         # Comment out the following line if you don't need perceptual loss
-        self.net = vgg16(pretrained=True, requires_grad=False)
+        # self.net = vgg16(pretrained=True, requires_grad=False)
```
Comment (PaParaZz1) [inline lzero/model/unizero_world_models/lpips.py:24]: why comment this part

File: lzero/model/unizero_world_models/kv_caching.py [hunk 4]
```diff
@@ -62,17 +62,17 @@ def get(self) -> torch.Tensor:
         """
         return self._cache[:, :, :self._size, :]
 
-    def update(self, x: torch.Tensor) -> None:
+    def update(self, x: torch.Tensor, tokens: int) -> None:
```
Comment (PaParaZz1) [inline lzero/model/unizero_world_models/kv_caching.py:65]: add comments about the new argument

File: zoo/atari/config/atari_rezero_mz_config.py [hunk 5]
```diff
@@ -19,6 +19,17 @@
 reuse_search = True
 collect_with_pure_policy = True
 buffer_reanalyze_freq = 1
+
```
Comment (PaParaZz1) [inline zoo/atari/config/atari_rezero_mz_config.py:22]: remove this

File: zoo/atari/config/atari_unizero_config.py [hunk 6]
```diff
@@ -1,105 +1,119 @@
...
-                max_blocks=num_unroll_steps,
-                max_tokens=2 * num_unroll_steps,  # NOTE: each timestep has 2 tokens: obs and action
-                context_length=2 * infer_context_length,
-                device='cuda',
+    # ====== only for debug =====
+    # collector_env_num = 8
+    # evaluator_env_num = 2
+    # num_simulations = 5
+    # max_env_step = int(2e5)
+    # reanalyze_ratio = 0.1
```
Comment (PaParaZz1) [inline zoo/atari/config/atari_unizero_config.py:27]: simplify unizero config

File: zoo/box2d/bipedalwalker/config/bipedalwalker_cont_sampled_unizero_config.py [hunk 7]
```diff
@@ -42,17 +42,17 @@
             world_model_cfg=dict(
                 obs_type='vector',
                 num_unroll_steps=num_unroll_steps,
-                policy_entropy_loss_weight=1e-4,
+                policy_entropy_loss_weight=5e-3,
                 continuous_action_space=continuous_action_space,
                 num_of_sampled_actions=K,
                 sigma_type='conditioned',
```
Comment (PaParaZz1) [inline zoo/box2d/bipedalwalker/config/bipedalwalker_cont_sampled_unizero_config.py:48]: why still conditioned here

File: zoo/dmc2gym/config/dmc2gym_pixels_suz_config.py [hunk 8]
```diff
@@ -116,7 +117,8 @@
         type='dmc2gym_lightzero',
         import_names=['zoo.dmc2gym.envs.dmc2gym_lightzero_env'],
     ),
-    env_manager=dict(type='subprocess'),
+    # env_manager=dict(type='subprocess'),
+    env_manager=dict(type='base'),
```
Comment (PaParaZz1) [inline zoo/dmc2gym/config/dmc2gym_pixels_suz_config.py:121]: why base here

File: lzero/model/unizero_world_models/lpips.py (hunk 3 above)
Comment (puyuan1996) [reply lzero/model/unizero_world_models/lpips.py:24]: to be more efficient (omit the model load process) when we don't use decode regularization.

File: zoo/box2d/bipedalwalker/config/bipedalwalker_cont_sampled_unizero_config.py (hunk 7 above)
Comment (puyuan1996) [reply zoo/box2d/bipedalwalker/config/bipedalwalker_cont_sampled_unizero_config.py:48]: We will rename it to policy_entropy_weight latter

File: lzero/entry/train_unizero.py (hunk 1 above)
Comment (puyuan1996) [reply lzero/entry/train_unizero.py:111]: polished now

File: zoo/atari/config/atari_rezero_mz_config.py [hunk 9]
```diff
@@ -33,6 +32,9 @@
         evaluator_env_num=evaluator_env_num,
         n_evaluator_episode=evaluator_env_num,
         manager=dict(shared_memory=False, ),
+        # # TODO: only for debug
+        # collect_max_episode_steps=int(20),
+        # eval_max_episode_steps=int(20),
```
Comment (PaParaZz1) [inline zoo/atari/config/atari_rezero_mz_config.py:37]: remove commented codes

File: lzero/worker/muzero_segment_collector.py [hunk 10]
```diff
@@ -0,0 +1,774 @@
...
+            env: BaseEnvManager = None,
+            policy: namedtuple = None,
+            tb_logger: 'SummaryWriter' = None,  # noqa
+            exp_name: Optional[str] = 'default_experiment',
+            instance_name: Optional[str] = 'collector',
+            policy_config: 'policy_config' = None,  # noqa
+    ) -> None:
+        """
+        Overview:
+            Initialize the MuZeroCollector with the given parameters.
```
Comment (PaParaZz1) [inline lzero/worker/muzero_segment_collector.py:46]: polish comments

File: lzero/worker/muzero_segment_collector.py [hunk 11]
```diff
@@ -0,0 +1,774 @@
...
+        last_game_segments[i].game_segment_to_array()
+
+        # put the game segment into the pool
+        self.game_segment_pool.append((last_game_segments[i], last_game_priorities[i], done[i]))
+
+        # reset last game_segments # TODO:origin
+        last_game_segments[i] = None
+        last_game_priorities[i] = None
+
+        return None
```
Comment (PaParaZz1) [inline lzero/worker/muzero_segment_collector.py:320]: remove this

File: lzero/worker/muzero_segment_collector.py [hunk 12]
```diff
@@ -0,0 +1,774 @@
...
+        if collect_with_pure_policy:
+            temp_visit_list = [0.0 for i in range(self._env.action_space.n)]
+
+        while True:
+            with self._timer:
+                # Get current ready env obs.
+                obs = self._env.ready_obs
+                ready_env_id = set(obs.keys())
+                if len(ready_env_id) < self._env_num:
+                    print(f'muzero_segment_collector: len(ready_env_id) < self._env_num, ready_env_id: {ready_env_id}')
```
Comment (PaParaZz1) [inline lzero/worker/muzero_segment_collector.py:419]: use logger

File: lzero/model/unizero_world_models/world_model.py [hunk 13]
```diff
@@ -87,7 +89,7 @@ def __init__(self, config: TransformerConfig, tokenizer) -> None:
             if self.sigma_type == 'fixed':
                 self.fixed_sigma_value = self.config.fixed_sigma_value
             else:
-                self.fixed_sigma_value = 0.3
+                self.fixed_sigma_value = 0.5
```
Comment (PaParaZz1) [inline lzero/model/unizero_world_models/world_model.py:92]: add an input argument to control it

File: lzero/model/unizero_world_models/world_model.py [hunk 14]
```diff
@@ -110,6 +112,138 @@ def __init__(self, config: TransformerConfig, tokenizer) -> None:
...
+        index = self.shared_pool_index_init_envs[env_id]
+        self.shared_pool_index_init_envs[env_id] = (self.shared_pool_index_init_envs[env_id] + 1) % self.shared_pool_size_init
+        
+        return index
+
+    #@profile
+    def custom_copy_kv_cache_to_shared_wm(self, src_kv: KeysValues) -> int:
+        """
+        Overview:
+            Efficiently copy the contents of a KeysValues object to the shared pool.
```
Comment (PaParaZz1) [inline lzero/model/unizero_world_models/world_model.py:179]: polish comments about similar methods

File: lzero/model/unizero_world_models/world_model.py [hunk 15]
```diff
@@ -353,6 +496,7 @@ def forward(self, obs_embeddings_or_act_tokens: Dict[str, Union[torch.Tensor, tu
         # logits_ends is None
         return WorldModelOutput(x, logits_observations, logits_rewards, None, logits_policy, logits_value)
 
+    #@profile
```
Comment (PaParaZz1) [inline lzero/model/unizero_world_models/world_model.py:499]: remove commented codes

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Pull request merged**